File: day_13_ruby/Advent_13_1.rb
require 'debug'
require 'json'
# file = File.open("dummy.txt")
file = File.open("data.txt")

file_data = file.read

file_data = file_data.split("\n")

file_data = file_data.select{|line| line.strip != ""}

puts "Parsing #{file_data.length/2} pairs."

def correct_order(pair)
    left = pair[0]
    right = pair[1]

    length = [pair[0].length, pair[1].length].max

    return nil if length <= 0

    (0...length).each do |i|
        if !left[i] && right[i]
            return true
        elsif left[i] && !right[i]
            return false
        end

        if left[i].class == Integer && right[i].class == Integer
            if left[i] < right[i]
                return true
            elsif left[i] > right[i]
                return false
            end

        elsif left[i].class == Array || right[i].class == Array
            result = correct_order(
                [left[i].class == Integer ? [left[i]] : left[i],
                right[i].class == Integer ? [right[i]] : right[i]]
            )

            return result if [true, false].include?(result)

        else
            return nil
        end
    end
end

line_number = 1
total = 0
current_pair = []

file_data.each do |str|
    line = JSON.parse(str.strip)
    current_pair.push(line)

    if line_number % 2 == 0
        valid = correct_order(current_pair)
        # debugger if valid.class == Range
        puts valid
        if valid
            total += line_number/2
        end
        current_pair = []
    end

    line_number += 1
end

puts total



# file_data.each_with_index do |line, y|
# end
# puts

File: day_13_ruby/Advent_13_1.js
"use strict"

function main(){

    const fs = require('fs');



    let file_data;

    // fs.readFile('data.txt', 'utf8', (err, data) => {
    // fs.readFile('dummy.txt', 'utf8', (err, data) => {
    // fs.readFile('small.txt', 'utf8', (err, data) => {
    fs.readFile('sorted.txt', 'utf8', (err, data) => {
        if(err) throw err;

        // file_data = data.split("\n");
        file_data = data.split("\r\n");
        console.log(`\nParsing ${file_data.length} lines.\n`);
        
        file_data = file_data.filter(line => line);
        console.log(`\nParsing ${file_data.length/2} pairs.\n`);
        
        function correct_order(current_pair){
            let left = current_pair[0]
            let right = current_pair[1]
            for(let i = 0; i < Math.max(left.length, right.length); i++){

                if(!left[i] && right[i]){
                    return true
                } else if (left[i] && !right[i]) {
                    return false
                }

                if(typeof(left[i]) == 'number' && typeof(right[i]) == 'number'){
                    if (left[i] < right[i]) {
                        return true
                        
                    } else  if(left[i] > right[i]) {
                        return false
                        
                    }
                } else if (
                    typeof(left[i]) == 'object' || typeof(right[i]) == 'object') {
                    const result = correct_order([
                        typeof(left[i]) == 'number' ? [left[i]] : left[i], 
                        typeof(right[i]) == 'number' ? [right[i]] : right[i]
                    ])

                    if(result == true || result == false){return result}

                }
            }
                
        }
    



        let line_number = 1;
        let total = 0;
        let current_pair = [];
       
        file_data.forEach((line) => {
            current_pair.push(JSON.parse(line));
            

            if(line_number % 2 == 0){
                
                let valid = correct_order(current_pair);

                console.log(`Reading pair ${line_number/2} as ${valid}`);

                if(valid){
                    total += line_number/2
                }
                current_pair = [];

            }
            
            line_number +=1;
        });

        console.log(total);
    })
}

main();